plugins/api/src/replica_close.cpp

```diff
@@ -0,0 +1,378 @@
+                log::api::trace("Replica Token = {}...", l1desc.replica_token);
+
+                auto entry = ix::replica_access_table::instance().erase_pid(l1desc.replica_token, getpid());
+
+                // Close the underlying file object.
+                if (const auto ec = close_physical_object(*_comm, l1desc.l3descInx); ec != 0) {
+                    if (entry) {
+                        ix::replica_access_table::instance().restore(*entry);
+                    }
+
+                    log::api::error("Failed to close file object [error_code={}].", ec);
+                    return ec;
+                }
+
+                log::api::trace("Releasing L1 descriptor ...");
+
+                if (const auto ec = freeL1desc(l1desc_index); ec != 0) {
+                    log::api::error("Failed to release L1 descriptor [error_code={}].", ec);
+                    return ec;
+                }
+            }
+
+            if (!json_input.contains("update_catalog") || json_input.at("update_catalog").get<bool>()) {
+                if (std::cend(json_input) != json_input.find("file_descriptor_info")) {
+                    log::api::trace("Updating the catalog ...");
+
+                    const auto& fd_info = json_input.at("file_descriptor_info");
+                    const auto size_on_disk = file_size(*_comm, fd_info);
+
+                    if (size_on_disk < 0) {
+                        log::api::error("Failed to retrieve the replica's size on disk [error_code={}].", size_on_disk);
+                        return size_on_disk;
+                    }
+
+                    const auto logical_path = fd_info.at("data_object_info").at("object_path").get<std::string>();
+                    const auto resource_hierarchy = fd_info.at("data_object_info").at("resource_hierarchy").get<std::string>();
+
+                    if (const auto ec = update_replica_size_and_status(*_comm, logical_path, resource_hierarchy, size_on_disk); ec != 0) {
+                        log::api::error("Failed to update the replica size in the catalog [error_code={}].", ec);
+                        return ec;
+                    }
+
+                    // Compute checksum and store it in the catalog.
+                    // TODO This could be optimized to avoid multiple lookups (use []).
+                    const auto compute_checksum = (!json_input.contains("compute_checksum") ||
+                                                    json_input.at("compute_checksum").get<bool>());
+
+                    if (compute_checksum) {
+                        log::api::trace("Computing checksum ...");
+                        const int replica_number = fd_info.at("data_object_info").at("replica_number").get<int>();
+                        constexpr const auto calculation = fs::verification_calculation::always;
+                        fs::server::data_object_checksum(*_comm, logical_path, replica_number, calculation);
+                    }
+                }
```

Comment:
this code feels too loose. it can be invoked with any JSON data following the correct structure.

Reply:
no longer an issue due to swapping of operations.